lib/src/main/java/org/esbtools/eventhandler/PromiseOfPromise.java

```diff
@@ -0,0 +1,69 @@
+/*
+ *  Copyright 2016 esbtools Contributors and/or its affiliates.
+ *
+ *  This file is part of esbtools.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.esbtools.eventhandler;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class PromiseOfPromise<U> implements Promise<U> {
```

Comment:
I thought maybe we could use the Guava strategy with an alternate type (AsyncFunction vs Function) for a callback which returns a Future, but this doesn't work with lambda expressions because they don't have type information. Spotify's FuturesExtra library recognized this same restriction, and so they have `asyncTransform` and `syncTransform`, analogous to `thenPromise` and `then` respectively. Again, we'd probably do well to adopt that family of terminology, but the concepts are basically the same.